pmd-java/src/main/java/net/sourceforge/pmd/lang/java/rule/design/NPathComplexityRule.java

```diff
@@ -4,262 +4,52 @@
-
-        int npath = npathStart;
-        JavaNode n;
-
-        for (int i = 0; i < node.jjtGetNumChildren(); i++) {
-            n = (JavaNode) node.jjtGetChild(i);
-            npath += (Integer) n.jjtAccept(this, data);
-        }
-
-        return npath;
-    }
-
-    @Override
-    public Object visit(ASTMethodDeclaration node, Object data) {
-        int npath = complexityMultipleOf(node, 1, data);
-
-        DataPoint point = new DataPoint();
-        point.setNode(node);
-        point.setScore(1.0 * npath);
-        point.setMessage(getMessage());
-        addDataPoint(point);
-
-        return Integer.valueOf(npath);
-    }
-
-    @Override
-    public Object visit(JavaNode node, Object data) {
-        int npath = complexityMultipleOf(node, 1, data);
-        return Integer.valueOf(npath);
-    }
-
-    @Override
-    public Object visit(ASTIfStatement node, Object data) {
-        // (npath of if + npath of else (or 1) + bool_comp of if) * npath of
-        // next
-
-        List<JavaNode> statementChildren = new ArrayList<>();
-        for (int i = 0; i < node.jjtGetNumChildren(); i++) {
-            if (node.jjtGetChild(i).getClass() == ASTStatement.class) {
-                statementChildren.add((JavaNode) node.jjtGetChild(i));
-            }
-        }
-
-        if (statementChildren.isEmpty() || statementChildren.size() == 1 && node.hasElse()
-                || statementChildren.size() != 1 && !node.hasElse()) {
-            throw new IllegalStateException("If node has wrong number of children");
-        }
-
-        // add path for not taking if
-        int complexity = 0;
-        if (!node.hasElse()) {
-            complexity++;
-        }
-
-        for (JavaNode element : statementChildren) {
-            complexity += (Integer) element.jjtAccept(this, data);
-        }
-
-        int boolCompIf = sumExpressionComplexity(node.getFirstChildOfType(ASTExpression.class));
-        return Integer.valueOf(boolCompIf + complexity);
-    }
-
-    @Override
-    public Object visit(ASTWhileStatement node, Object data) {
-        // (npath of while + bool_comp of while + 1) * npath of next
-
-        int boolCompWhile = sumExpressionComplexity(node.getFirstChildOfType(ASTExpression.class));
-
-        Integer nPathWhile = (Integer) ((JavaNode) node.getFirstChildOfType(ASTStatement.class)).jjtAccept(this, data);
+public class NPathComplexityRule extends AbstractJavaMetricsRule {
 
-        return Integer.valueOf(boolCompWhile + nPathWhile + 1);
-    }
-
-    @Override
-    public Object visit(ASTDoStatement node, Object data) {
-        // (npath of do + bool_comp of do + 1) * npath of next
-
-        int boolCompDo = sumExpressionComplexity(node.getFirstChildOfType(ASTExpression.class));
-
-        Integer nPathDo = (Integer) ((JavaNode) node.getFirstChildOfType(ASTStatement.class)).jjtAccept(this, data);
+    private static final IntegerProperty REPORT_LEVEL_DESCRIPTOR = new IntegerProperty(
```

Comment:
We have now a different property (well, the name is different). Assume, someone uses the rule in a custom rule, such as:

``` 
    <rule ref="java-codesize/NPathComplexity">
        <properties>
            <property name="minimum" value="100.0" />
        </properties>
    </rule>
```

We would now fail executing the ruleset with the error "Cannot set non-existant property 'minimum' on Rule NPathComplexity".

We could add backwards compatibility, by defining the old property additionally, e.g.:

`private static final DoubleProperty MINIMUM_DESCRIPTOR = new DoubleProperty("minimum", "Deprecated! Minimum reporting threshold", 0d, 300d, 200d,
            2.0f);`

And get the actual value by this way:

```
    private int getReportLevel() {
        double oldProp = getProperty(MINIMUM_DESCRIPTOR);
        if (oldProp != 200.0) {
            System.err.println("Warning! Rule NPathComplexity uses deprecated property 'minimum'. Future versions of PMD will remove support for this property. Please use 'reportLevel' instead!");
            return (int)oldProp;
        }

        return getProperty(REPORT_LEVEL_DESCRIPTOR);
    }
```

Is this worth it? The user might now update to 6.0.0 and it "just works" - not sure, if the warning is ever noticed... and with 7.0.0 it suddenly fails.... Maybe we just move the failure one release into the future :)

I've just tried it, it's at least possible...

In any case, I'll write a note into the release notes, that the rules has been revamped and the property is different now, before I'll merge it.

@jsotuyod Do you think, such a backwards-compatibility is necessary?

Reply:
I love and totally support this idea!

Reply:
Making it compatible and logging a warning is probably the smoothest we can do so +1